File: android/keyboard/src/main/kotlin/com/viettelex/keyboard/DisplayCase.kt
package com.viettelex.keyboard

/**
 * Case chuẩn cho proper noun khi HIỂN THỊ gợi ý (port iOS). Chỉ token không nhập nhằng;
 * token nhập nhằng viết hoa THEO CHUỖI ([chains]).
 */
object DisplayCase {
    fun apply(w: String, after: String? = null): String {
        val p = after?.lowercase()
        if (p != null && chains[p]?.contains(w) == true) return Cp.capitalizeFirst(w)
        return proper[w] ?: w
    }

    val chains: Map<String, Set<String>> = hashMapOf(
        // địa danh VN
        "hà" to setOf("nội", "nam", "tĩnh", "giang"),
        "đà" to setOf("nẵng", "lạt"),
        "nha" to setOf("trang"),
        "phú" to setOf("quốc", "thọ", "yên"),
        "sài" to setOf("gòn"),
        "hồ" to setOf("chí"),
        "chí" to setOf("minh"),
        "việt" to setOf("nam"),
        "cần" to setOf("thơ"),
        "hải" to setOf("phòng", "dương"),
        "hạ" to setOf("long"),
        "quy" to setOf("nhơn"),
        "vũng" to setOf("tàu"),
        "biên" to setOf("hòa"),
        "bình" to setOf("dương", "định", "thuận"),
        "thanh" to setOf("hóa"),
        "hội" to setOf("an"),
        "quảng" to setOf("ninh", "nam", "ngãi", "bình", "trị"),
        "nghệ" to setOf("an"),
        "lâm" to setOf("đồng"),
        "đồng" to setOf("nai", "tháp"),
        "long" to setOf("an"),
        "tây" to setOf("ninh"),
        "new" to setOf("york"),
        "san" to setOf("francisco"),
        "hong" to setOf("kong"),
        // họ → đệm (chuỗi tên người)
        "nguyễn" to setOf("văn", "thị", "đức", "minh", "ngọc", "hữu", "xuân", "thùy", "kim", "hồng", "quốc", "đình"),
        "trần" to setOf("văn", "thị", "đức", "minh", "ngọc", "quốc"),
        "lê" to setOf("văn", "thị", "đức", "minh", "ngọc", "hữu"),
        "phạm" to setOf("văn", "thị", "minh", "ngọc"),
        "hoàng" to setOf("văn", "thị", "minh", "anh"),
        "vũ" to setOf("văn", "thị", "minh"),
        "đặng" to setOf("văn", "thị"),
        "bùi" to setOf("văn", "thị"),
        "đỗ" to setOf("văn", "thị"),
        "ngô" to setOf("văn", "thị"),
        "dương" to setOf("văn", "thị"),
        "trịnh" to setOf("văn", "thị", "xuân"),
        // đệm → tên
        "văn" to setOf("hùng", "tuấn", "dũng", "sơn", "nam", "long", "hải", "minh"),
        "thị" to setOf("hương", "lan", "thu", "ngọc", "hồng", "phương", "hà", "linh"),
    )

    val proper: Map<String, String> = hashMapOf(
        // thương hiệu / sản phẩm
        "senprints" to "SenPrints", "printik" to "Printik",
        "apple" to "Apple", "iphone" to "iPhone", "ipad" to "iPad", "macbook" to "MacBook",
        "samsung" to "Samsung", "google" to "Google", "facebook" to "Facebook",
        "youtube" to "YouTube", "tiktok" to "TikTok", "zalo" to "Zalo",
        "shopee" to "Shopee", "lazada" to "Lazada", "grab" to "Grab",
        "momo" to "MoMo", "vnpay" to "VNPay", "vietcombank" to "Vietcombank",
        "techcombank" to "Techcombank", "viettel" to "Viettel", "vingroup" to "Vingroup",
        "vinfast" to "VinFast", "vietjet" to "Vietjet", "fpt" to "FPT",
        "arsenal" to "Arsenal", "liverpool" to "Liverpool", "chelsea" to "Chelsea",
        "manchester" to "Manchester", "bitcoin" to "Bitcoin", "ethereum" to "Ethereum",
        "binance" to "Binance", "netflix" to "Netflix", "spotify" to "Spotify",
        "github" to "GitHub", "claude" to "Claude", "chatgpt" to "ChatGPT",
        "tesla" to "Tesla", "openai" to "OpenAI", "anthropic" to "Anthropic",
        "yahoo" to "Yahoo", "hotmail" to "Hotmail",
        // OS / phần mềm
        "windows" to "Windows", "macos" to "macOS", "ios" to "iOS",
        "android" to "Android", "linux" to "Linux", "chrome" to "Chrome",
        "safari" to "Safari", "excel" to "Excel", "word" to "Word",
        "wifi" to "WiFi", "gmail" to "Gmail", "outlook" to "Outlook",
        // địa danh nước ngoài (token đơn, không nhập nhằng)
        "singapore" to "Singapore", "dubai" to "Dubai", "london" to "London",
        "tokyo" to "Tokyo", "sydney" to "Sydney", "canada" to "Canada",
        "houston" to "Houston", "seattle" to "Seattle", "miami" to "Miami",
        "dallas" to "Dallas", "austin" to "Austin", "texas" to "Texas",
        "california" to "California", "washington" to "Washington",
        "york" to "York", "francisco" to "Francisco", "paris" to "Paris",
        "bangkok" to "Bangkok", "seoul" to "Seoul",
        // họ Việt an toàn (gần như luôn là tên riêng khi đứng một mình).
        // CÁC TOKEN NHẬP NHẰNG bị loại có chủ ý: vũ (vũ khí), đỗ (đỗ xe),
        // ngô (bắp ngô), dương (đại dương), trang (trang web), nội (nội bộ),
        // quốc (tổ quốc) — hiện thường an toàn hơn hoa sai.
        "nguyễn" to "Nguyễn", "trịnh" to "Trịnh", "đặng" to "Đặng", "bùi" to "Bùi",
    )
}

File: iOS/Keyboard/DisplayCase.swift
// DisplayCase — case chuẩn cho proper noun khi HIỂN THỊ gợi ý. Datastore
// case-fold toàn bộ để đếm ("senprints"), nhưng nút gợi ý và chữ được chèn
// phải ra "SenPrints". CHỈ chứa token không-nhập-nhằng: brand/sản phẩm/OS,
// địa danh nước ngoài, họ Việt an toàn. Token nhập nhằng (văn ↔ văn bản,
// pháp ↔ phương pháp, trần ↔ trần nhà) KHÔNG đưa vào — thà hiện thường còn
// hơn viết hoa sai giữa câu.
import Foundation

enum DisplayCase {
    /// Trả về dạng hiển thị chuẩn; từ không có trong bảng giữ nguyên.
    /// `after`: từ đứng ngay trước — token nhập nhằng được viết hoa THEO NGỮ
    /// CẢNH chuỗi tên riêng ("hà"→"nội" hiện "Nội", nhưng "nội" đơn lẻ giữ
    /// thường vì còn là "nội bộ").
    static func apply(_ w: String, after prev: String? = nil) -> String {
        if let p = prev?.lowercased(), chains[p]?.contains(w) == true {
            return w.prefix(1).uppercased() + w.dropFirst()
        }
        return proper[w] ?? w
    }

    /// Chuỗi tên riêng: prev → các next chỉ-proper-khi-đứng-sau-prev.
    static let chains: [String: Set<String>] = [
        // địa danh VN
        "hà": ["nội", "nam", "tĩnh", "giang"],
        "đà": ["nẵng", "lạt"],
        "nha": ["trang"],
        "phú": ["quốc", "thọ", "yên"],
        "sài": ["gòn"],
        "hồ": ["chí"], "chí": ["minh"],
        "việt": ["nam"],
        "cần": ["thơ"],
        "hải": ["phòng", "dương"],
        "hạ": ["long"],
        "quy": ["nhơn"],
        "vũng": ["tàu"],
        "biên": ["hòa"],
        "bình": ["dương", "định", "thuận"],
        "thanh": ["hóa"],
        "hội": ["an"],
        "quảng": ["ninh", "nam", "ngãi", "bình", "trị"],
        "nghệ": ["an"],
        "lâm": ["đồng"],
        "đồng": ["nai", "tháp"],
        "long": ["an"],
        "tây": ["ninh"],
        "new": ["york"], "san": ["francisco"], "hong": ["kong"],
        // họ → đệm (chuỗi tên người)
        "nguyễn": ["văn", "thị", "đức", "minh", "ngọc", "hữu", "xuân", "thùy", "kim", "hồng", "quốc", "đình"],
        "trần": ["văn", "thị", "đức", "minh", "ngọc", "quốc"],
        "lê": ["văn", "thị", "đức", "minh", "ngọc", "hữu"],
        "phạm": ["văn", "thị", "minh", "ngọc"],
        "hoàng": ["văn", "thị", "minh", "anh"],
        "vũ": ["văn", "thị", "minh"],
        "đặng": ["văn", "thị"],
        "bùi": ["văn", "thị"],
        "đỗ": ["văn", "thị"],
        "ngô": ["văn", "thị"],
        "dương": ["văn", "thị"],
        "trịnh": ["văn", "thị", "xuân"],
        // đệm → tên
        "văn": ["hùng", "tuấn", "dũng", "sơn", "nam", "long", "hải", "minh"],
        "thị": ["hương", "lan", "thu", "ngọc", "hồng", "phương", "hà", "linh"],
    ]

    static let proper: [String: String] = [
        // thương hiệu / sản phẩm
        "senprints": "SenPrints", "printik": "Printik",
        "apple": "Apple", "iphone": "iPhone", "ipad": "iPad", "macbook": "MacBook",
        "samsung": "Samsung", "google": "Google", "facebook": "Facebook",
        "youtube": "YouTube", "tiktok": "TikTok", "zalo": "Zalo",
        "shopee": "Shopee", "lazada": "Lazada", "grab": "Grab",
        "momo": "MoMo", "vnpay": "VNPay", "vietcombank": "Vietcombank",
        "techcombank": "Techcombank", "viettel": "Viettel", "vingroup": "Vingroup",
        "vinfast": "VinFast", "vietjet": "Vietjet", "fpt": "FPT",
        "arsenal": "Arsenal", "liverpool": "Liverpool", "chelsea": "Chelsea",
        "manchester": "Manchester", "bitcoin": "Bitcoin", "ethereum": "Ethereum",
        "binance": "Binance", "netflix": "Netflix", "spotify": "Spotify",
        "github": "GitHub", "claude": "Claude", "chatgpt": "ChatGPT",
        "tesla": "Tesla", "openai": "OpenAI", "anthropic": "Anthropic",
        "yahoo": "Yahoo", "hotmail": "Hotmail",
        // OS / phần mềm
        "windows": "Windows", "macos": "macOS", "ios": "iOS",
        "android": "Android", "linux": "Linux", "chrome": "Chrome",
        "safari": "Safari", "excel": "Excel", "word": "Word",
        "wifi": "WiFi", "gmail": "Gmail", "outlook": "Outlook",
        // địa danh nước ngoài (token đơn, không nhập nhằng)
        "singapore": "Singapore", "dubai": "Dubai", "london": "London",
        "tokyo": "Tokyo", "sydney": "Sydney", "canada": "Canada",
        "houston": "Houston", "seattle": "Seattle", "miami": "Miami",
        "dallas": "Dallas", "austin": "Austin", "texas": "Texas",
        "california": "California", "washington": "Washington",
        "york": "York", "francisco": "Francisco", "paris": "Paris",
        "bangkok": "Bangkok", "seoul": "Seoul",
        // họ Việt an toàn (gần như luôn là tên riêng khi đứng một mình).
        // CÁC TOKEN NHẬP NHẰNG bị loại có chủ ý: vũ (vũ khí), đỗ (đỗ xe),
        // ngô (bắp ngô), dương (đại dương), trang (trang web), nội (nội bộ),
        // quốc (tổ quốc) — hiện thường an toàn hơn hoa sai.
        "nguyễn": "Nguyễn", "trịnh": "Trịnh", "đặng": "Đặng", "bùi": "Bùi",
    ]
}
